tests/bsim/bluetooth/host/l2cap/credits_seg_recv/src/main.c

```diff
@@ -0,0 +1,450 @@
+/*
+ * Copyright (c) 2022 Nordic Semiconductor
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include "bstests.h"
+#include "common.h"
+
+#define LOG_MODULE_NAME main
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(LOG_MODULE_NAME, LOG_LEVEL_DBG);
+
+CREATE_FLAG(is_connected);
+CREATE_FLAG(flag_l2cap_connected);
+
+#define L2CAP_MPS CONFIG_BT_L2CAP_TX_MTU
+#define SDU_NUM   3
+#define SDU_LEN   (2 * L2CAP_MPS)
+/* We intentionally send smaller SDUs than the channel can fit */
+#define L2CAP_MTU (2 * SDU_LEN)
+
+/* Only one SDU transmitted or received at a time */
+NET_BUF_POOL_DEFINE(sdu_pool, 1, L2CAP_MTU, 8, NULL);
+
+static uint8_t tx_data[SDU_LEN];
+static uint16_t rx_cnt;
+
+K_SEM_DEFINE(sdu_received, 0, 1);
+
+struct test_ctx {
+	struct bt_l2cap_le_chan le_chan;
+	size_t tx_left;
+} test_ctx;
+
+int l2cap_chan_send(struct bt_l2cap_chan *chan, uint8_t *data, size_t len)
+{
+	LOG_DBG("chan %p conn %u data %p len %d", chan, bt_conn_index(chan->conn), data, len);
+
+	struct net_buf *buf = net_buf_alloc(&sdu_pool, K_NO_WAIT);
+
+	if (buf == NULL) {
+		FAIL("No more memory\n");
+		return -ENOMEM;
+	}
+
+	net_buf_reserve(buf, BT_L2CAP_SDU_CHAN_SEND_RESERVE);
+	net_buf_add_mem(buf, data, len);
+
+	int ret = bt_l2cap_chan_send(chan, buf);
+
+	ASSERT(ret >= 0, "Failed sending: err %d", ret);
+
+	LOG_DBG("sent %d len %d", ret, len);
+	return ret;
+}
+
+struct net_buf *alloc_buf_cb(struct bt_l2cap_chan *chan)
+{
+	return net_buf_alloc(&sdu_pool, K_NO_WAIT);
+}
+
+void continue_sending(struct test_ctx *ctx)
+{
+	struct bt_l2cap_chan *chan = &ctx->le_chan.chan;
+
+	LOG_DBG("%p, left %d", chan, ctx->tx_left);
+
+	if (ctx->tx_left) {
+		l2cap_chan_send(chan, tx_data, sizeof(tx_data));
+	} else {
+		LOG_DBG("Done sending %u", bt_conn_index(chan->conn));
+	}
+}
+
+void sent_cb(struct bt_l2cap_chan *chan)
+{
+	LOG_DBG("%p", chan);
+
+	if (test_ctx.tx_left) {
+		test_ctx.tx_left--;
+	}
+
+	continue_sending(&test_ctx);
+}
+
+void recv_cb(struct bt_l2cap_chan *l2cap_chan, uint16_t sdu_len, uint16_t seg_offset,
+	     uint16_t seg_len, uint8_t *seg)
+{
+	LOG_DBG("sdu len %u frag offset %u frag len %u", sdu_len, seg_offset, seg_len);
+
+	ASSERT(sdu_len == sizeof(tx_data), "Recv SDU length does not match send length.");
+
+	/* Verify SDU data matches TX'd data. */
+	ASSERT(memcmp(seg, &tx_data[seg_offset], seg_len) == 0, "RX data doesn't match TX");
```

Comment:
Suggest to use `FAIL` instead of `ASSERT`, as `ASSERT`s can be disabled (either by purpose or accident)